volatility3/framework/plugins/windows/suspicious_threads.py

```diff
@@ -0,0 +1,201 @@
+
+        if vad_path is None:
+            # set this so checks after report the non file backed region in the path column
+            vad_path = "<Non-File Backed Region>"
+
+            yield (
+                vad_path,
+                "This thread started execution in the VAD starting at base address ({:#x}), which is not backed by a file".format(
+                    vad_base
+                ),
+            )
+
+        if prot != "PAGE_EXECUTE_WRITECOPY":
+            yield (
+                vad_path,
+                "VAD at base address ({:#x}) hosting this thread has an unexpected starting protection {}".format(
+                    vad_base, prot
+                ),
+            )
+
+        if (
+            exe_path
+            and vad_path.lower().endswith(".exe")
+            and (vad_path.lower() != exe_path.lower())
+        ):
+            yield (
+                vad_path,
+                "VAD at base address ({:#x}) hosting this thread maps an application executable that is not the process exectuable".format(
+                    vad_base
+                ),
+            )
+
+    def _enumerate_processes(self, kernel, all_ranges):
+        filter_func = pslist.PsList.create_pid_filter(self.config.get("pid", None))
+
+        for proc in pslist.PsList.list_processes(
+            context=self.context,
+            layer_name=kernel.layer_name,
+            symbol_table=kernel.symbol_table_name,
+            filter_func=filter_func,
+        ):
+            ranges = self._get_ranges(kernel, all_ranges, proc)
+
+            # smeared vads or process is terminating
+            if len(all_ranges[proc.vol.offset]) < 5:
+                continue
+
+            pid = proc.UniqueProcessId
+            proc_name = utility.array_to_string(proc.ImageFileName)
+
+            _, __, exe_path = self._get_range(ranges, proc.SectionBaseAddress)
+            if not isinstance(exe_path, str):
+                exe_path = None
+
+            yield proc, pid, proc_name, exe_path, ranges
+
+    def _generator(self):
+        kernel = self.context.modules[self.config["kernel"]]
+
+        all_ranges = {}
+
+        for proc, pid, proc_name, exe_path, ranges in self._enumerate_processes(
+            kernel, all_ranges
+        ):
+            # processes often schedule multiple threads at the same address
+            # there is no benefit to checking the same address more than once per process
+            checked = set()
+
+            for thread in threads.Threads.list_threads(kernel, proc):
+                # do not process if a thread is exited or terminated (4 = Terminated)
+                if thread.ExitTime.QuadPart > 0 or thread.Tcb.State == 4:
+                    continue
+
+                # bail if accessing the threads members causes a page fault
+                info = thrdscan.ThrdScan.gather_thread_info(thread)
+                if not info:
+                    continue
+
+                tid, start_address = info[2], info[3]
```

Comment:
Direct accessing numeric components of a return value is a little...  nasty.  Any chance we could either return a `NamedTuple` or use `_, _, tid, start_address, _, _ ... = ...`.  It just means that changes are more evident, so that if the order or something about the format of the functions return changes, this will still work or throw up an error quickly...

Reply:
done